internal/gensample/gensample.go

```diff
@@ -0,0 +1,737 @@
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if gapicFname == "" {
+			errChan <- nil
+			return
+		}
+		f, err := os.Open(gapicFname)
+		if err != nil {
+			errChan <- errors.E(err, "cannot read GAPIC config file: %q", gapicFname)
+			return
+		}
+		defer f.Close()
+
+		if err := yaml.NewDecoder(f).Decode(&gen.gapic); err != nil {
+			errChan <- errors.E(err, "error reading GAPIC config file: %q", gapicFname)
+			return
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		errChan <- gen.readSampleConfigFiles(sampleFnames)
+	}()
+
+	gen.nofmt = nofmt
+	if err := <-errChan; err != nil {
+		return nil, err
+	}
+	if err := <-errChan; err != nil {
+		return nil, err
+	}
+	wg.Wait()
+	return &gen, nil
+}
+
+// GenMethodSamples generators samples from protos and configurations stored in the generator,
+// and writes the generated samples to generator.Outputs.
+func (gen *generator) GenMethodSamples() error {
+	gen.disambiguateSampleIDs()
+	gen.Outputs = make(map[string][]byte)
+
+	for _, samp := range gen.sampleConfig.Samples {
+		var iface GAPICInterface
+		var method GAPICMethod
+		for _, iface = range gen.gapic.Interfaces {
+			if iface.Name == samp.Service {
+				for _, method = range iface.Methods {
+					if method.Name == samp.Rpc {
+						break
+					}
+				}
+				break
+			}
+		}
+
+		if method.Name != samp.Rpc {
+			return errors.E(nil, "generating sample %q: rpc %q not found", samp.ID, samp.Rpc)
+		}
+		if iface.Name != samp.Service {
+			return errors.E(nil, "generating sample %q: service %q not found", samp.ID, samp.Service)
+		}
+
+		gen.reset()
+		if err := gen.genSample(*samp, method); err != nil {
+			err = errors.E(err, "generating: %s.%s:%s", iface.Name, method.Name, samp.ID)
+			log.Fatal(err)
+		}
+
+		content, err := gen.commit(!gen.nofmt, time.Now().Year())
+		if err != nil {
+			return err
+		}
+
+		fname := samp.ID + ".go"
+		gen.Outputs[fname] = content
+	}
+	return nil
+}
+
+type generator struct {
+	desc         []*descriptor.FileDescriptorProto
+	descInfo     pbinfo.Info
+	gapic        GAPICConfig
+	sampleConfig schema_v1p2.SampleConfig
+	clientPkg    pbinfo.ImportSpec
+	nofmt        bool
+
+	pt      printer.P
+	imports map[pbinfo.ImportSpec]bool
+	Outputs map[string][]byte
+}
+
+// readSampleConfigFiles loads sample configs from local files and store
+// them in generator.sampleConfig.
+func (gen *generator) readSampleConfigFiles(paths []string) error {
+	for _, path := range paths {
+		f, err := os.Open(path)
+		if err != nil {
+			return errors.E(err, "cannot open sample config file: %s", path)
+		}
+		decoder := yaml.NewDecoder(f)
+		config, err := readSampleConfig(decoder, path)
+		if err != nil {
+			return err
+		}
+		gen.sampleConfig.Samples = append(gen.sampleConfig.Samples, config.Samples...)
+	}
+	return nil
+}
+
+func (g *generator) reset() {
+	g.pt.Reset()
+	for imp := range g.imports {
+		delete(g.imports, imp)
+	}
+}
+
+func (g *generator) commit(gofmt bool, year int) ([]byte, error) {
```

Comment:
I didn't write this function but I actually think this is a great name, because this method does is exactly to commit what is around in the buffer inside `generator` to bytes.

But I agree this is a bit confusing for folks who haven't been working closely with this part of the code. I'll add documentation to list out what `commit` does.